### PR TITLE
feat: add support for optional dynamic route params

### PR DESCRIPTION
**Pull request by mbhrznr**
tackles #1690.

this pr is a first attempt to support optional dynamic routes.  
the file name convention is inspired by [sveltekit](https://kit.svelte.dev/docs/advanced-routing)'s established convention.

the feature can be tested in the primary fixture by running:
```
deno task fixture
```

once started, navigate to `/i18n/lang` and `/i18n/:lang/lang`.

### CODE REVIEW

**Comment by mbhrznr** [comment]
would add some documentation once we've settled on the general approach.

File: src/server/context_test.ts [hunk 1]
```diff
@@ -21,11 +21,20 @@ Deno.test("pathToPattern", async (t) => {
...
 
+  await t.step("parses optional params", () => {
+    assertEquals(pathToPattern("foo/[[name]]"), "/foo{/:name}?");
+    assertEquals(pathToPattern("foo/[name]/[[bob]]"), "/foo/:name{/:bob}?");
+    assertEquals(pathToPattern("foo/[[name]]/bar"), "/foo{/:name}?/bar");
+    assertEquals(
+      pathToPattern("foo/[[name]]/bar/[[bob]]"),
+      "/foo{/:name}?/bar{/:bob}?",
+    );
+  });
```
Comment (marvinhagemeister) [inline src/server/context_test.ts:36]: Let's add one test case for invalid usage like:

```sh
foo/foo-[[name]]-bar/baz
foo/[[name]]-bar/baz
foo/foo-[[name]]/baz
foo/foo-[[name]]
foo/[[name]]-bar
```

Currently, that returns this path `/foo{/:name}?-bar/baz` which would match `foo-bar/baz`. To me that shouldn't be allowed. When using the double bracket notation `[[asd]]` it should be only possible if it's a full segment of the path.

I think adding a check a double `[[` is not at the start of a segment or a `]]` is not at the end of a segment should suffice.

**Review by marvinhagemeister — COMMENTED** [review]
This is fantastic! Going with the way Svelte does it makes a lot of sense. Like this much more then the original proposed `[...id]`. Found one additional case to account for, but other than that it looks really good and I'm excited to have this in Fresh 👍

File: src/server/context_test.ts (hunk 1 above)
Comment (mbhrznr) [reply src/server/context_test.ts:36]: thanks for writing down the tests cases, this made it incredible easy to add!

**Comment by mbhrznr** [comment]
would it make sense to refactor `www`'s `docs/[...slug].tsx` to make use of the new concept?
or would you prefer to have this tackled in another pr?

**Comment by marvinhagemeister** [comment]
> would it make sense to refactor www's docs/[...slug].tsx to make use of the new concept?
or would you prefer to have this tackled in another pr?

Let's tackle that in another PR

**Review by marvinhagemeister — APPROVED** [review]
Awesome, this is great!

**Review by marvinhagemeister — APPROVED** [review]
Awesome, this is great!

**Pull request merged**